### PR TITLE
[macOS]Support SemanticsService.announce

### PR DESCRIPTION
**Pull request by hangyujin**
issue:https://github.com/flutter/flutter/issues/99715

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm [hunk 1]
```diff
@@ -967,7 +986,22 @@ - (void)onAccessibilityStatusChanged:(NSNotification*)notification {
...
+
+    NSAccessibilityPriorityLevel priority = [assertiveness isEqualToNumber:@1]
+                                                ? NSAccessibilityPriorityHigh
+                                                : NSAccessibilityPriorityMedium;
 
+    NSAccessibilityPostNotificationWithUserInfo(
+        [self viewControllerForId:kFlutterDefaultViewId].flutterView,
+        NSAccessibilityAnnouncementRequestedNotification,
+        @{NSAccessibilityAnnouncementKey : message, NSAccessibilityPriorityKey : @(priority)});
+  }
```
Comment (cbracken) [inline shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:1003]: Top-level AppKit API functions are always a bit tricky to test. There are a few tricks you can use for this. A simple one in this case is:
1. Extract this to a method like:
```objc
- (void)announceAccessibilityMessage:(NSString*)withPriority:(NSAccessibilityPrioirityKey);
```
2. Use `CreateMockFlutterEngine()` to create a partial mock in your test.
3. Verify the method got called in the test:
```objc
[[engineMock expect] announceAccessibilityMessage:@"Foo" withPriority:...];

// Fire an event here.

[engineMock verify];
```

**Review by cbracken — COMMENTED** [review]
A couple minor nits, otherwise looking good!

**Review by cbracken — APPROVED** [review]
Looks good! Thanks!

File: shell/platform/darwin/macos/framework/Source/FlutterEngineTest.mm [hunk 2]
```diff
@@ -766,6 +766,27 @@ - (nonnull NSView*)createWithViewIdentifier:(int64_t)viewId arguments:(nullable
   EXPECT_TRUE(triedToTerminate);
 }
 
+TEST_F(FlutterEngineTest, HandleAccessibilityEvent) {
+  bool announced = false;
+  id engineMock = CreateMockFlutterEngine(nil);
+
+  OCMStub([engineMock announceAccessibilityMessage:[OCMArg any] withPriority:[OCMArg any]])
+      .andDo((^(NSInvocation* invocation) {
+        announced = true;
```
[the remark this answers is not in the document]
Comment (cbracken) [reply shell/platform/darwin/macos/framework/Source/FlutterEngineTest.mm:775]: +1. You can get these via `[invocation getArgument:AtIndex:]`.

File: shell/platform/darwin/macos/framework/Source/FlutterEngine.mm [hunk 3]
```diff
@@ -967,7 +986,30 @@ - (void)onAccessibilityStatusChanged:(NSNotification*)notification {
 
   self.semanticsEnabled = enabled;
 }
+- (void)handleAccessibilityEvent:(NSDictionary<NSString*, id>*)annotatedEvent {
+  NSString* type = annotatedEvent[@"type"];
+  if ([type isEqualToString:@"announce"]) {
+    NSString* message = annotatedEvent[@"data"][@"message"];
+    NSNumber* assertiveness = annotatedEvent[@"data"][@"assertiveness"];
```
Comment (loic-sharma) [inline shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:993]: Does this crash if `data`, `message`, or `assertiveness` properties are missing? A malformed message should be ignored (though an error should be logged ideally)

Comment (cbracken) [reply shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:993]: These will result in `nil`.

Comment (loic-sharma) [reply shell/platform/darwin/macos/framework/Source/FlutterEngine.mm:993]: Ah OK perfect! :)

**Review by chunhtai — APPROVED** [review]
LGTM

**Pull request merged**